packages/react-dom/src/__tests__/ReactDOMServerIntegrationHooks-test.internal.js

```diff
@@ -843,4 +853,384 @@ describe('ReactDOMServerHooks', () => {
+          <div>
+            <div
+              aria-labelledby="c_0"
+            />
+            <div
+              id="c_0"
+            />
+          </div>
+      `);
+    });
+
+    it('useOpaqueReference: IDs match when, after hydration, a new component that uses the ID is rendered', async () => {
+      let _setShowDiv;
+      function App() {
+        const id = useOpaqueReference();
+        const [showDiv, setShowDiv] = useState(false);
+        _setShowDiv = setShowDiv;
+
+        return (
+          <div>
+            <div id={id}>Child One</div>
+            {showDiv && <div id={id}>Child Two</div>}
+          </div>
+        );
+      }
+
+      const container = document.createElement('div');
+      document.body.append(container);
+
+      container.innerHTML = ReactDOMServer.renderToString(<App />);
+      ReactDOM.hydrate(<App />, container);
+      expect(container).toMatchInlineSnapshot(`
+        <div>
+          <div
+            data-reactroot=""
+          >
+            <div
+              id="s_0"
+            >
+              Child One
+            </div>
+          </div>
+        </div>
+      `);
+      await ReactTestUtils.act(async () => {
+        _setShowDiv(true);
+      });
+      expect(container).toMatchInlineSnapshot(`
+        <div>
+          <div
+            data-reactroot=""
+          >
+            <div
+              id="c_0"
+            >
+              Child One
+            </div>
+            <div
+              id="c_0"
+            >
+              Child Two
+            </div>
+          </div>
+        </div>
+      `);
+    });
+
+    if (__EXPERIMENTAL__) {
+      it('useOpaqueReference: IDs match when part of the DOM tree is server rendered and part is client rendered', async () => {
+        let suspend = false;
+        let resolve;
+        let promise = new Promise(resolvePromise => (resolve = resolvePromise));
+
+        function Child({text}) {
+          if (suspend) {
+            throw promise;
+          } else {
+            return text;
+          }
+        }
+
+        function RenderedChild() {
+          useEffect(() => {
+            Scheduler.unstable_yieldValue('Child did commit');
+          });
+          return null;
+        }
+
+        function App() {
+          const id = useOpaqueReference();
+          useEffect(() => {
+            Scheduler.unstable_yieldValue('Did commit');
+          });
+          return (
+            <div>
+              <div id={id}>Child One</div>
+              <RenderedChild />
+              <React.Suspense fallback={'Fallback'}>
+                <div id={id}>
+                  <Child text="Child Two" />
+                </div>
+              </React.Suspense>
+            </div>
+          );
+        }
+
+        const container = document.createElement('div');
+
+        container.innerHTML =
+          '<div data-reactroot=""><div id="s_0">Child One</div><!--$!--><div>Fallback</div><!--/$--></div>';
```

Comment:
I'd use `ReactDOMServer.renderToString(...)` to render this string so that you don't have to hard code the format.

Example: https://github.com/facebook/react/blob/9fe1031244903e442de179821f1d383a9f2a59f2/packages/react-dom/src/__tests__/ReactDOMServerPartialHydration-test.internal.js#L1049